src/SIC/INet.hs

```diff
@@ -0,0 +1,118 @@
+module SIC.INet where
+
+import qualified Data.Map.Strict as M
+import qualified Data.Set        as Set
+import           Protolude       hiding (reduce)
+
+import           SIC.Type
+
+makeNet ∷ [Node] → Net
+makeNet nodes =
+  let n = M.fromList $ zip [0..] nodes
+  in Net n (findRedexes n)
+
+findRedexes ∷ M.Map Int Node → [(Int, Int)]
+findRedexes nodes =
+  go Set.empty $ M.toList $ (toNode . slot P) <$> nodes
+  where
+    isFree n = maybe False (\x -> kind x == Free) $ M.lookup n nodes
+    go s ((a, b):xs)
+      | isFree a && a == b = go (Set.insert (a, b) s) xs
+      | isFree a || isFree b || Set.member (b, a) s = go s xs
+      | otherwise = go (Set.insert (a, b) s) xs
```

Comment:
Should we check here that the port on the connecting node is primary?

It looks like this will include primary-auxiliary connections as redexes.

Reply:
Yeah we should. I had some logic to do that earlier but forgot to add it back when I added the `Free` nodes, since we want `Free - Free` to be a redex when it's a self-edge.

Reply:
Oh, wait that doesn't work either. Stand by...